tools/workflow/cleaner/lambda/clean.go

```diff
@@ -15,48 +15,129 @@ const Type = "lambda"
 				logger.Printf("Try to delete function %s created-at %s", *lf.FunctionArn, *lf.LastModified)
 				deleteFunctionIDs = append(deleteFunctionIDs, lf.FunctionArn)
 			}
 		}
-		if ListFunctionsOutput.NextMarker == nil {
+		if listFunctionsOutput.NextMarker == nil {
 			break
 		}
-		nextToken = ListFunctionsOutput.NextMarker
+		nextToken = listFunctionsOutput.NextMarker
 	}
 	if len(deleteFunctionIDs) < 1 {
 		logger.Printf("No Lambda functions to delete")
-		return nil
+	} else {
+		for _, id := range deleteFunctionIDs {
+			terminateFunctionInput := &lambda.DeleteFunctionInput{FunctionName: id}
+			if _, err := lambdaClient.DeleteFunction(terminateFunctionInput); err != nil {
+				return fmt.Errorf("unable to delete function: %w", err)
+			}
+		}
+		logger.Printf("Deleted %d Lambda functions", len(deleteFunctionIDs))
+	}
+
+	return nil
+}
+
+func cleanLayers(sess *session.Session, expirationDate time.Time) error {
+	lambdaClient := lambda.New(sess)
+	var deleteLayerVersions []Layer
+	var nextToken *string
+
+	for {
+		listLayerVersionsInput := &lambda.ListLayersInput{
+			Marker: nextToken,
+		}
+
+		// Retrieve layer versions from Lambda service
+		listLayerVersionsOutput, err := lambdaClient.ListLayers(listLayerVersionsInput)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve layer versions: %w", err)
+		}
+
+		// Loop through retrieved layer versions
+		for _, layer := range listLayerVersionsOutput.Layers {
+			layerARN := shouldDeleteLayer(layer)
+			if layerARN != nil {
+				created, err := time.Parse("2006-01-02T15:04:05Z0700", *layer.LatestMatchingVersion.CreatedDate)
+				if err != nil {
+					return fmt.Errorf("error parsing layer created time: %w", err)
+				}
+				if expirationDate.After(created) {
+					logger.Printf("Try to delete layer version %s created-at %s", *layer.LayerArn, created)
+					deleteLayerVersions = append(deleteLayerVersions, Layer{layerARN, layer.LatestMatchingVersion.Version})
+				}
+			}
+		}
+
+		if listLayerVersionsOutput.NextMarker == nil {
+			break
+		}
+		nextToken = listLayerVersionsOutput.NextMarker
 	}
 
-	for _, id := range deleteFunctionIDs {
-		terminateFunctionInput := &lambda.DeleteFunctionInput{FunctionName: id}
-		if _, err := lambdaclient.DeleteFunction(terminateFunctionInput); err != nil {
-			return fmt.Errorf("unable to delete function: %w", err)
+	if len(deleteLayerVersions) < 1 {
+		logger.Printf("No Lambda layers to delete")
+	} else {
+		for _, id := range deleteLayerVersions {
+			for *id.Version > 0 {
+				// Prepare input for deleting a specific layer version
+				deleteLayerVersionInput := &lambda.DeleteLayerVersionInput{
+					LayerName:     id.LayerARN,
+					VersionNumber: id.Version,
+				}
+				if _, err := lambdaClient.DeleteLayerVersion(deleteLayerVersionInput); err != nil {
+					return fmt.Errorf("unable to delete layer version: %w", err)
+				}
+				// Decrement the version number for the next iteration
+				*id.Version--
```

Comment:
Idk the likelihood of it happening, but you are assuming that all these layer versions exist. What if they don't?

Reply:
Usually, the layer version is increased only when a layer is pushed with same name, its very unlikely that we will miss a version number. unless someone manually delete the versions between. I think we should be okay since we dont play much around integ account.